KEYCLOAK-6262 Incorporate new visual design from PatternFly

junezhang opened the pull request:
Currently, UXD team finished the new login design in PatternFly. http://www.patternfly.org/pattern-library/application-framework/login-page/

### This PR is for
-  incorporating the new visual design from PatternFly to KeyCloak.

### The screenshots for the new login of KeyCloak

#### Basic Login Page (no social provider)
<img width="200" alt="screen shot 2018-02-07 at 6 30 08 pm" src="https://user-images.githubusercontent.com/701009/35912079-84d143a8-0c36-11e8-890c-370efa090dfb.png">

Responsive Design (on mobile)
<img width="200" alt="screen shot 2018-02-07 at 6 32 37 pm" src="https://user-images.githubusercontent.com/701009/35912105-9533a1c8-0c36-11e8-9cbb-f277ba9560b8.png">


#### Login with social provider
One column (when the amount is less or equal 4)
<img width="200" alt="screen shot 2018-02-07 at 6 35 54 pm" src="https://user-images.githubusercontent.com/701009/35912122-a30f521a-0c36-11e8-8e3f-68b3b27e1de9.png">


Two columns (when the amount is more than 4)
<img width="200" alt="screen shot 2018-02-07 at 6 34 59 pm" src="https://user-images.githubusercontent.com/701009/35912164-bab15e18-0c36-11e8-9f0d-49403f1db306.png">

Responsive Design (on mobile)
<img width="100" alt="screen shot 2018-02-07 at 6 36 05 pm" src="https://user-images.githubusercontent.com/701009/35912178-c833433a-0c36-11e8-8eba-029c716c790b.png">

A gif to check what the new login looks like
<img width="200" src="https://user-images.githubusercontent.com/701009/35965548-2dec7ef2-0cf6-11e8-80a9-2bc3bfdb0b83.gif">

Commit:
KEYCLOAK-6262 Incorporate new visual design from PatternFly

Update the username or email

Fix narrow/wide in template.ftl

minor style update

Add the Realm HTML name and image

Config OTP and Update Password
Not display the locale selector if there is less than 1 locale.

Fix margins/paddings on config otp screens

Fix title

Upgraded to PatternFly 3.41.6

Added RCUE and updated RH-SSO login theme

Refine the RCUE padding issue

Fix tests

Fix Keycloak background

Fix

stianst commented on themes/src/main/resources/theme/base/login/template.ftl:
```diff
@@ -32,62 +32,55 @@
+    <header class="${properties.kcHeaderClass!}">
+      <div id="kc-logo"><a href="${properties.kcLogoLink!'#'}"><div id="kc-logo-wrapper"></div></a></div>
+    </header>
+    <div class="${properties.kcFormCardClass!} <#if realm.password && social.providers??>${properties.kcFormCardAccountClass!}</#if>">
```
This is causing a problem on some pages as "realm.password" and "social.providers" are not available on all pages. I don't quite understand what this is doing and why it adds this class like this. Can you explain?

junezhang replied:
Sure, @stianst. The logical of this piece of code is the login box width is different, when no social provide, the width is 500px. and when adding the social provider the width of login box is 840px.
As here, I need a class to control the width.

stianst replied:
https://github.com/junezhang/keycloak/pull/1

stianst reviewed (dismissed):
Found a few broken things:

* Login page shows "Username" when it should show "Username or email"
* Forgot password page shows "Username or email" above input instead of inside
* Title on all pages is "Login In" (password reset page, register page, update password page, etc.)
* Grant page is broken
* Configure OTP page is broken
* OTP page is broken
* IdP Link to existing account is broken
* Terms and conditions page is broken
* Error page is completely broken due to "if realm.password .." added to template.ftl

junezhang commented:
@stianst Updating as following:

1, Update the username placeholder to 
`<#if !realm.loginWithEmailAllowed>${msg("username")}<#elseif !realm.registrationEmailAsUsername>${msg("usernameOrEmail")}<#else>${msg("email")}</#if>`
2, Update the Forgot password page shows "Username or email" above input instead of inside.
same with the above method. 
3, Update the Grant Page
<img width="300" alt="screen shot 2018-02-12 at 8 28 51 pm" src="https://user-images.githubusercontent.com/701009/36097809-70c1d634-1037-11e8-9795-8ab48f67e8f2.png">
4, Update the Term Page
<img width="300" alt="screen shot 2018-02-12 at 8 28 45 pm" src="https://user-images.githubusercontent.com/701009/36097827-8934c50a-1037-11e8-90e1-bacf611cd507.png">

As in my local server, when I open "Update password" "Config OTP", it leads to an error page.
The server error log is here
https://gist.github.com/junezhang/205a03c286e1464ab1ee3624679a1ef5
https://gist.github.com/anonymous/918e1f5e1eb4f8aaf6a4b16c71a80047

Please help to check it.
Thanks.

stianst commented:
One more thing we need to consider. We have the Realm HTML name which allows users to show a name or a image for the realm. This is now completely gone and the realm name doesn't exist anywhere. That means a user doesn't know what realm they are authenticating.

junezhang commented:
@stianst Have fixed the Realm name and image as the attachment.

Realm with image
<img width="300" src="https://user-images.githubusercontent.com/701009/36712599-80b1f6e4-1bc4-11e8-9f40-d2dbee95eecc.png">

Realm with text
<img width="300" src="https://user-images.githubusercontent.com/701009/36712607-86864782-1bc4-11e8-8a1e-6014698d725a.png">

junezhang commented:
@stianst @ssilvert 
I've fixed the "Config OTP", "Update Password", "Update profile".

<img width="200px" src="https://user-images.githubusercontent.com/701009/36974362-5f594de8-20b1-11e8-8450-8047a3d9ee24.png" >

<img width="200px" src="https://user-images.githubusercontent.com/701009/36974369-641b9f20-20b1-11e8-8757-1462b9ee7908.png" >

<img width="200px" src="https://user-images.githubusercontent.com/701009/36974371-69568f72-20b1-11e8-9ad2-4d5837d6036a.png" >

stianst commented:
Fixed some minor margin/padding issue in otp config screens:
https://github.com/keycloak/keycloak/pull/4983/commits/e36dfcc1406e0d382ddf2912700bfcdbe3a7cd5d

stianst commented:
There's two outstanding issues:

* Title is "Log In" on all pages. This doesn't make any sense. We should probably change this to have the realm name on the top always (above the white box) and replace "Log In" with a page specific title.

* Weird title "MY REALM ACCOUNT." on consent screen. This is broken in master as well. The fix for the above would help here. Keep "MY REALM" at the top and replace "Log In" with something like "Grant access to App Name".

stianst reviewed (changes requested):
There's two outstanding issues:

* Title is "Log In" on all pages. This doesn't make any sense. We should probably change this to have the realm name on the top always (above the white box) and replace "Log In" with a page specific title.

* Weird title "MY REALM ACCOUNT." on consent screen. This is broken in master as well. The fix for the above would help here. Keep "MY REALM" at the top and replace "Log In" with something like "Grant access to App Name".

stianst commented:
We need to review RH-SSO theme and maybe also update RCUE styles

stianst reviewed (changes requested):
We need to review RH-SSO theme and maybe also update RCUE styles

stianst commented:
![image](https://user-images.githubusercontent.com/2271511/37350462-2659e9c0-26d9-11e8-92a7-935d24dd7850.png)

RH-SSO theme is completely broken, so that needs looking into.

junezhang commented:
@stianst The RH-SSO theme look like below:
<img width="300" alt="screen shot 2018-03-14 at 6 12 25 pm" src="https://user-images.githubusercontent.com/701009/37399629-23a16c46-27bd-11e8-846e-dfa0c91ac9ec.png">
How do you think if we add the Red Hat logo on the top-right?

stianst commented:
Why did you add themes/src/main/resources-product/theme/rh-sso/login/resources/css/login.css? I deliberately removed it and added login-rhsso.css instead so it can inherit styles from Keycloak theme instead of duplicating it. We've had quite a few issues here in the past where the two have been inconsitent.

junezhang commented:
@stianst Removed the login.css from resources-product. please have a review.

seb54000 commented:
I think the minus sign in the login.css file for display higher than 768px is a caps
https://github.com/keycloak/keycloak/blob/190771ddf18d89a3242737d533a01b48bec82556/themes/src/main/resources-product/theme/rh-sso/login/resources/css/login.css#L272 

This lead to a behaviour of getting locale listbox (for internationalization) out of the screen

junezhang commented:
@seb54000 Thanks for your feedback.

In our latest login design, we fixed this issue. 
https://github.com/junezhang/keycloak/blob/c8e6a9947e28a3a2385dea9bf41b11f201658867/themes/src/main/resources/theme/keycloak/login/resources/css/login.css

The screenshot is as below
<img width="300"  src="https://user-images.githubusercontent.com/701009/37455841-09c01694-2879-11e8-96ce-45153802563c.png">

mobile screenshot
<img width="200" alt="screen shot 2018-03-15 at 5 51 27 pm" src="https://user-images.githubusercontent.com/701009/37456092-9002227e-2879-11e8-9aa0-2724a3c7be0d.png">

Please help to check. thanks.

stianst commented:
A lot of tests are failing. This is primarily due to the changes in the page title. I'm currently working on updating the tests and will send a commit when I have tests passing.

seb54000 commented:
@junezhang Great ! :raised_hands:

stianst commented:
@junezhang squashed the commits into a single commit and rebased from master so it should now be ready to merge once @vmuzikar has reviewed

vmuzikar commented:
I've tried to thoroughly test the new login screen in all supported browsers (both mobile and desktop) and found a few issues (their priority/severity is probably a subject to discuss).

**Overflowing text for some social provider buttons**
This is happening when the browser window is a bit narrow but not enough for switching to the mobile version. This is mainly a problem for tablets in portrait mode.
Also IdPs with long names don't have a proper padding.
<img width="483" alt="screen shot 2018-03-16 at 11 23 10" src="https://user-images.githubusercontent.com/13906855/37518133-324eb26e-2914-11e8-9b6a-7fa6bc554690.png">

**Overflowing realm name**
Long realm names are overflowing and covering localization dropdown on smartphones.
![screen shot 2018-03-16 at 11 27 05](https://user-images.githubusercontent.com/13906855/37525099-16e9429c-292c-11e8-9ef2-f7e6831691f6.png)

**Placeholders don't work while focused in IE**
When you select a text input field which has a placeholder (the labeling text), the placeholder disappears even though the field is blank (i.e. no text was entered yet).
This could be an UX problem mainly in combination with autofocus which we use and could the user confuse a bit. Take the following 2 scenarios:
- User opens the login page and sees two fields from which one is just blank (with no hint if they should enter username, email, ...).
- User clicks "Forgot Password?" on the login page and sees just one blank input field with no clue what should they enter.
![capture](https://user-images.githubusercontent.com/13906855/37525988-99aa65ec-292e-11e8-9e0b-85d3103a16e2.PNG)

**Console UI Tests are broken**
Some of them are failing in IE and FF, all of them in Chrome.
I can fix them but it won't be sooner than Monday (and possibly even Tuesday).

@stianst WDYT? How severe do you consider the issues? Should we merge it now or should we at least fix the tests? I'd prefer the latter.

Commit:
fix the overflowing issue

junezhang commented:
@vmuzikar Thanks for checking.

Fixed the overflowing text for some social provider buttons
![screen shot 2018-03-19 at 4 04 34 pm](https://user-images.githubusercontent.com/701009/37584183-72213c54-2b8f-11e8-8ac7-db70d8d3854c.png)

Fixed overflowing realm name
![screen shot 2018-03-19 at 4 05 25 pm](https://user-images.githubusercontent.com/701009/37584218-9a9d07b2-2b8f-11e8-8a31-370574197048.png)

Currently, in my local server, I don't have IE, so can't test it.

vmuzikar commented:
@junezhang Thanks for the fixes!

I believe I've discovered another one.
On Android phones it seems like the IdP buttons which have no icon (e.g. GitLab) are very "grayed out" and even a different font (larger size?).
![screen shot 2018-03-19 at 12 34 30](https://user-images.githubusercontent.com/13906855/37593757-69415694-2b73-11e8-907b-349612ad6a7f.png)
However, it might be an issue specific to the Android Emulator I use. We need to try to reproduce it with a real Android phone.

vmuzikar committed:
Fix Console UI Tests to reflect the new login page

vmuzikar commented:
@junezhang I've fixed the Console UI Tests and created a PR for it: https://github.com/junezhang/keycloak/pull/2

vmuzikar commented:
Also, I can confirm that the issue described in previous comment (https://github.com/keycloak/keycloak/pull/4983#issuecomment-374186101) is present with real Android phones as well.

Commit:
Merge pull request #2 from vmuzikar/new-login-page

Fix Console UI Tests to reflect the new login page

Commit:
Fix the different style of the IdP buttons
Fix the IE placeholder issue - add label

junezhang commented:
@vmuzikar fixed the different style in different Idp buttons. Below is the screenshots of the emulated device - Nexus. please help to check on your Android phone, if it is right?
![screen shot 2018-03-20 at 12 25 15 pm](https://user-images.githubusercontent.com/701009/37635968-fc7487a8-2c39-11e8-9e25-cca475c4823a.png)

vmuzikar commented:
@junezhang Thanks for the fixes.

I can confirm that the IdP buttons now look good on Android phones.

However, I'm not sure about adding the labels back. Imho it doesn't look quite "elegant" having both placeholders and labels. @stianst WDYT?

![screen shot 2018-03-20 at 10 16 40](https://user-images.githubusercontent.com/13906855/37646832-ffbbaba8-2c2a-11e8-9774-bd172e476640.png)

I see 3 options here.
1. Use labels instead placeholders in the IE solely
1. Use some different approach for placeholders which would achieve their full functionality in the IE. That would, however, require Javascript and that makes it not a usable solution for us, I suppose.
1. Use only the placeholders and just ignore a bit worse UX in the IE.

Also, we need to fix the issue with the StackOverflow button on tablets - https://github.com/keycloak/keycloak/pull/4983#issuecomment-374181605.

vmuzikar commented:
@stianst Should we include some Red Hat logo in the RH-SSO theme?

stianst commented:
@vmuzikar there's a separate JIRA for updating the logo

stianst commented:
I would remove the placeholders completely. We don't want to introduce JS in the login frames and as such there's no way to properly support IE. It also makes it consistent with other pages.

@junezhang I believe you may have just forgotten to remove the placeholders after re-introducing the labels?

vmuzikar commented:
@stianst +1
Even though the placeholders looked very cool, clean and modern (I really liked them), labels are the only solution how to provide seamless and functional design across all the browsers.

stianst commented:
We can probably revisit labels later

stianst committed:
Removed placeholder on login and reset pass. Fixed Keycloak background on wide screens.

stianst commented:
Removed the placeholders. Also fixed the Keycloak background which wasn't showing properly on wide-screens.

Commit:
fixed the stackoverflow issues
fixed the width in the tablets
